cachier/pickle_core.py

```diff
@@ -131,52 +131,113 @@ def _get_cache(self):
                 self._reload_cache()
             return self.cache
 
-    def _save_cache(self, cache):
+    def _get_cache_by_key(self, key=None, hash=None):
+        fpath = self._cache_fpath()
+        if hash is None:
+            fpath += f'_{hashlib.sha256(pickle.dumps(key)).hexdigest()}'
+        else:
+            fpath += f'_{hash}'
+        try:
+            with portalocker.Lock(fpath, mode='rb') as cache_file:
+                try:
+                    res = pickle.load(cache_file)
+                except EOFError:
+                    res = None
+        except FileNotFoundError:
+            res = None
+        return res
+
+    def _clear_all_cache_files(self):
+        fpath = self._cache_fpath()
+        path, name = os.path.split(fpath)
+        for subpath in os.listdir(path):
+            if subpath.startswith(name):
+                os.remove(os.path.join(path, subpath))
+
+    def _clear_being_calculated_all_cache_files(self):
+        fpath = self._cache_fpath()
+        path, name = os.path.split(fpath)
+        for subpath in os.listdir(path):
+            if subpath.startswith(name):
+                entry = self._get_cache_by_key(hash=subpath.split('_')[-1])
+                entry['being_calculated'] = False
+                self._save_cache(entry, hash=subpath.split('_')[-1])
+
+    def _save_cache(self, cache, key=None, hash=None):
         with self.lock:
             self.cache = cache
             fpath = self._cache_fpath()
+            if key is not None:
+                fpath += f'_{hashlib.sha256(pickle.dumps(key)).hexdigest()}'
+            elif hash is not None:
+                fpath += f'_{hash}'
             with portalocker.Lock(fpath, mode='wb') as cache_file:
                 pickle.dump(cache, cache_file, protocol=4)
-            self._reload_cache()
+            if key is None:
+                self._reload_cache()
 
     def get_entry_by_key(self, key, reload=False):  # pylint: disable=W0221
         with self.lock:
-            # print('{}, {}'.format(self.reload, reload))
-            if self.reload or reload:
-                self._reload_cache()
-            return key, self._get_cache().get(key, None)
+            if self.separate_files:
+                return key, self._get_cache_by_key(key)
```

Comment:
Is it ok that `separate_files` caches ignore the `reload` keyword? I get that this type of cache saves each entry to a separate file (to the extent of hash collisions, I guess), so does this invalidates the use of `reload`? Because `reload` was meant to make the caches thread-safe, addressing the fact that without per-read cache reloading different threads will have different versions of the cache. Or at least that what's my documentation of the keyword says (honestly can't remember). :)

Reply:
I don't think it's a problem because without `separate_files`, the pickle files are opened on each call.